### PR TITLE
Implement a taglib for selecting a file for a blobSoftRef

### PR DESCRIPTION
**Pull request by jmuscireum**
Bildschirmfoto 2020-10-15 um 13 11 00

Bildschirmfoto 2020-10-15 um 13 11 48

Bildschirmfoto 2020-10-15 um 13 12 05

Bildschirmfoto 2020-10-15 um 13 22 11

Bildschirmfoto 2020-10-15 um 13 12 35

### CODE REVIEW

**Review by idlira — APPROVED** [review]

File: src/main/java/sirius/biz/storage/layer2/BlobController.java [hunk 1]
```diff
@@ -86,4 +85,39 @@ public void uploadFile(final WebContext ctx,
...
     }
+
+    /**
+     * Find the {@link Blob} for a given path and returns info about it.
+     *
+     * @param webContext the request to handle
+     * @param out        the response to the AJAX call
+     * @param spaceName  the {@link BlobStorageSpace} to find the object in
+     */
+    @Routed(value = "/dasd/blob-info-for-path/:1", jsonCall = true)
```
Comment (meg1502) [inline src/main/java/sirius/biz/storage/layer2/BlobController.java:96]: what does the prefix 'dasd' stand for?

Comment (jmuscireum) [reply src/main/java/sirius/biz/storage/layer2/BlobController.java:96]: Direct Attached Storage Device

File: src/main/resources/default/taglib/w/blobSoftRefField.html.pasta [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+
+<script type="text/javascript">
+    $(document).ready(function () {
+        const $element = $('#___id');
+        const $blobKeyField = $('#___inputId');
+        const blobStorageSpace = '___blobRef.getSpace()';
+        const url = '___blobRef.url().asDownload().buildURL().orElse("")';
+        const filename = '___blobRef.getFilename()';
+        const path = '___blobRef.getPath()';
+        const defaultPreview = '___defaultPreview';
```
Comment (meg1502) [inline src/main/resources/default/taglib/w/blobSoftRefField.html.pasta:57]: if you just call one function with this, pretty pointless to define these const IMO

**Review by meg1502 — CHANGES_REQUESTED** [review]
would be nice to reduce amount of jquery used (getting rid of all is probably not possible when trying to use the bootstrap popover function)

File: src/main/java/sirius/biz/storage/layer2/BlobController.java [hunk 3]
```diff
@@ -86,4 +85,39 @@ public void uploadFile(final WebContext ctx,
                             .handle();
         }
     }
+
+    /**
+     * Find the {@link Blob} for a given path and returns info about it.
```
Comment (bwiedmann) [inline src/main/java/sirius/biz/storage/layer2/BlobController.java:90]:
```suggestion
     * Finds the {@link Blob} for a given path and returns info about it.
```

File: src/main/resources/default/assets/javascript/blobreffield.js [hunk 4]
```diff
@@ -0,0 +1,237 @@
+/*
+ * Made with all the love in the world
+ * by scireum in Remshalden, Germany
+ *
+ * Copyright by scireum GmbH
+ * http://www.scireum.de - info@scireum.de
+ */
+
+/**
+ * Initialize the given blobSoftRefField
```
Comment (bwiedmann) [inline src/main/resources/default/assets/javascript/blobreffield.js:10]:
```suggestion
 * Initializes the given blobSoftRefField
```

**Review by bwiedmann — CHANGES_REQUESTED** [review]
Typos

File: src/main/resources/default/taglib/w/blobSoftRefField.html.pasta (hunk 2 above)
Comment (sabieber) [reply src/main/resources/default/taglib/w/blobSoftRefField.html.pasta:57]: Do you suggest inlining them? If not const is good codestyle

Comment (jmuscireum) [reply src/main/resources/default/taglib/w/blobSoftRefField.html.pasta:57]: I'd prefer to leave it like this. Inlining just creates an illegible mess.

File: src/main/resources/default/assets/javascript/blobreffield.js [hunk 5]
```diff
@@ -43,81 +43,89 @@ function initBlobSoftRefField($element,
...
-    const $urlButton = $element.find('[data-toggle=popover]');
-    const $resetButton = $element.find('.btn-reset-js');
-    const $fileElement = $element.find('.file-js');
+    const selectButton = element.querySelector('.btn-select-file-js');
+    const urlButton = element.querySelector('[data-toggle=popover]');
+    const resetButton = element.querySelector('.btn-reset-js');
+    const fileElement = element.querySelector('.file-js');
 
-    $element.data('path', originalPath);
+    element.setAttribute('data-path', originalPath);
```
Comment (sabieber) [inline src/main/resources/default/assets/javascript/blobreffield.js]:
```suggestion
    element.dataset.path = originalPath;
```

File: src/main/resources/default/assets/javascript/blobreffield.js [hunk 6]
```diff
@@ -43,81 +43,89 @@ function initBlobSoftRefField($element,
...
+    $(urlButton).on('inserted.bs.popover', function () {
+        const popover = urlButton.parentElement.querySelector('.popover');
+        const closeButton = popover.querySelector('.button-close');
+        const applyButton = popover.querySelector('.button-apply');
+        const input = popover.querySelector('input');
 
-        $closeButton.click(function () {
-            $popup.popover('hide');
+        closeButton.addEventListener('click', function (e) {
+            e.preventDefault();
```
Comment (sabieber) [inline src/main/resources/default/assets/javascript/blobreffield.js:57]:
```suggestion
        closeButton.addEventListener('click', function (event) {
            event.preventDefault();
```

File: src/main/resources/default/assets/javascript/blobreffield.js [hunk 7]
```diff
@@ -43,81 +43,89 @@ function initBlobSoftRefField($element,
...
+            checkURL(input.value, applyButton);
         });
 
-        $input.keyup(function (e) {
+        input.addEventListener('paste', function () {
+            checkURL(input.value, applyButton);
+        });
+
+        input.addEventListener('keyup', function (e) {
             if (e.which === 13) {
```
Comment (sabieber) [inline src/main/resources/default/assets/javascript/blobreffield.js]:
```suggestion
        input.addEventListener('keyup', function (event) {
            if (event.key === 'Enter') {
```

File: src/main/resources/default/assets/javascript/blobreffield.js [hunk 8]
```diff
@@ -43,81 +43,89 @@ function initBlobSoftRefField($element,
...
-                updateURL($input.val());
+                updateURL(input.value);
             }
         });
 
-        $applyButton.click(function () {
-            updateURL($input.val());
+        applyButton.addEventListener('click', function (e) {
+            e.preventDefault();
```
Comment (sabieber) [inline src/main/resources/default/assets/javascript/blobreffield.js:76]:
```suggestion
        applyButton.addEventListener('click', function (event) {
            event.preventDefault();
```

File: src/main/resources/default/assets/javascript/blobreffield.js [hunk 9]
```diff
@@ -43,81 +43,89 @@ function initBlobSoftRefField($element,
...
+        input.select();
 
-        const $input = $(this).next('.popover').find('input');
-        const $applyButton = $(this).next('.popover').find('.button-apply');
+        checkURL(input.value, applyButton);
+    });
 
-        $input.val(url).select();
+    urlButton.addEventListener('click', function (e) {
+        e.preventDefault();
```
Comment (sabieber) [inline src/main/resources/default/assets/javascript/blobreffield.js:93]:
```suggestion
    urlButton.addEventListener('click', function (event) {
        event.preventDefault();
```

File: src/main/resources/default/assets/javascript/blobreffield.js [hunk 10]
```diff
@@ -43,81 +43,89 @@ function initBlobSoftRefField($element,
...
+        urlButton.blur();
 
-        checkURL($input.val(), $applyButton);
+        $(urlButton).popover('toggle');
     });
 
-    $resetButton.click(function (e) {
+    resetButton.addEventListener('click', function (e) {
         e.preventDefault();
```
Comment (sabieber) [inline src/main/resources/default/assets/javascript/blobreffield.js]:
```suggestion
    resetButton.addEventListener('click', function (event) {
        event.preventDefault();
```

File: src/main/resources/default/assets/javascript/blobreffield.js [hunk 11]
```diff
@@ -43,81 +43,89 @@ function initBlobSoftRefField($element,
...
+        blobKeyField.value = '';
 
         updateFile('');
         updateResetButton();
     });
 
-    $selectButton.click(function () {
-        const currentPath = $element.data('path') || blobStorageSpacePath;
+    selectButton.addEventListener('click', function (e) {
+        e.preventDefault();
```
Comment (sabieber) [inline src/main/resources/default/assets/javascript/blobreffield.js:110]:
```suggestion
    selectButton.addEventListener('click', function (event) {
        event.preventDefault();
```

File: src/main/resources/default/assets/javascript/blobreffield.js [hunk 12]
```diff
@@ -43,81 +43,89 @@ function initBlobSoftRefField($element,
...
+
+        const currentPath = element.getAttribute('data-path') || blobStorageSpacePath;
```
Comment (sabieber) [inline src/main/resources/default/assets/javascript/blobreffield.js:112]:
```suggestion
        const currentPath = element.dataset.path || blobStorageSpacePath;
```

File: src/main/resources/default/assets/javascript/blobreffield.js [hunk 13]
```diff
@@ -43,81 +43,89 @@ function initBlobSoftRefField($element,
...
 
         selectVFSFile(currentPath, blobStorageSpacePath).then(function (selectedValue) {
             $.getJSON('/dasd/blob-info-for-path/' + blobStorageSpace, {
                 path: selectedValue.substring(blobStorageSpacePath.length)
             }, function (json) {
-                $blobKeyField.val(json.fileId);
-                $element.data('path', selectedValue);
+                blobKeyField.value = json.fileId;
+                element.setAttribute('data-path', selectedValue);
```
Comment (sabieber) [inline src/main/resources/default/assets/javascript/blobreffield.js]:
```suggestion
                element.dataset.path = selectedValue;
```

File: src/main/resources/default/assets/javascript/blobreffield.js [hunk 14]
```diff
@@ -241,5 +232,14 @@ function initBlobSoftRefField(element,
...
-    updateFile(originalUrl, originalFilename);
+    const determineFilename = function (path) {
+        return path.substr(path.lastIndexOf("/") + 1);
+    };
+
+    const determineParentPath = function (path) {
+        return path.substr(0, path.lastIndexOf("/"));
+    };
+
+    element.setAttribute('data-path', determineParentPath(originalPath));
```
Comment (sabieber) [inline src/main/resources/default/assets/javascript/blobreffield.js:243]:
```suggestion
    element.dataset.path = determineParentPath(originalPath);
```

**Review by sabieber — CHANGES_REQUESTED** [review]

**Review by meg1502 — APPROVED** [review]

**Review by sabieber — APPROVED** [review]

File: src/main/java/sirius/biz/storage/layer2/BlobHardRef.java [hunk 15]
```diff
@@ -95,6 +95,19 @@ public String getFilename() {
         return getBlob().getFilename();
     }
 
+    /**
+     * Determines the path of the referenced blob.
+     *
+     * @return the path, or <tt>null</tt> if no blob is referenced
+     */
+    public String getPath() {
```
Comment (andyHa) [inline src/main/java/sirius/biz/storage/layer2/BlobHardRef.java:103]: @Nullable

File: src/main/resources/default/assets/javascript/blobreffield.js [hunk 16]
```diff
@@ -0,0 +1,245 @@
+/*
```
Comment (andyHa) [inline src/main/resources/default/assets/javascript/blobreffield.js:1]: sexy! 👍

**Review by andyHa — APPROVED** [review]

**Review by idlira — APPROVED** [review]

**Review by bwiedmann — APPROVED** [review]

**Review by sabieber — APPROVED** [review]

**Review by andyHa — APPROVED** [review]

**Pull request merged**